modeshape-jcr/src/main/java/org/modeshape/jcr/cache/document/WritableSessionCache.java

```diff
@@ -144,11 +147,19 @@ public WritableSessionCache( ExecutionContext context,
         this.referrerChangesForRemovedNodes = new HashMap<NodeKey, ReferrerChanges>();
         this.binaryReferencesByNodeKey = new ConcurrentHashMap<NodeKey, Set<BinaryKey>>();
         this.txns = repositoryEnvironment.getTransactions();
+        this.nodeTypes = repositoryEnvironment.nodeTypes();
     }
 
     protected final void assertInSession( SessionNode node ) {
         assert this.changedNodes.get(node.getKey()) == node : "Node " + node.getKey() + " is not in this session";
     }
+    
+    protected NodeTypes nodeTypes() {
+        if (this.nodeTypes == null && workspaceCache().repositoryEnvironment() != null) {
+            this.nodeTypes = workspaceCache().repositoryEnvironment().nodeTypes();
+        }  
```

Comment:
This class is `@ThreadSafe` and the `nodeTypes` field is not marked as `volatile`. However, I *think* that might be okay since the actual `NodeTypes` reference returned by `workspaceCache().repositoryEnvironment().nodeTypes()` should always be the same instance. In that case, this logic is idempotent, and its okay for the field to not be `volatile`.